zebra/connected.c

```diff
@@ -474,6 +480,60 @@ void connected_down(struct interface *ifp, struct connected *ifc)
 	rib_delete(afi, SAFI_MULTICAST, zvrf->vrf->vrf_id, ZEBRA_ROUTE_CONNECT,
 		   0, 0, &p, NULL, &nh, 0, zvrf->table_id, 0, 0, false);
 
+	/* When the last IPv4 address of an interface is deleted, Linux removes
+	 * all routes using this interface without any Netlink advertisement.
+	 * The removed routes include those that only have this particular
+	 * interface as a nexthop. Among those, remove the kernel one from the
+	 * FRR RIB and reinstall the other that have been added from FRR.
+	 */
+	if (afi == AFI_IP && count_ipv4 == 0 && if_is_operative(ifp)) {
+		RB_FOREACH (vrf, vrf_id_head, &vrfs_by_id) {
+			zvrf_iter = vrf->info;
+
+			if (!zvrf_iter)
+				continue;
+
+			table = zvrf_iter->table[AFI_IP][SAFI_UNICAST];
+			if (!table)
+				continue;
+
+			for (rn = route_top(table); rn;
```

Comment:
Is it not possible to look for affected nexthops through NHT? Scanning the entire RIB is quite costly.

Reply:
NHT table is void in the following case
```
r1# show ip route
Codes: K - kernel route, C - connected, S - static, R - RIP,
       O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, D - SHARP,
       F - PBR, f - OpenFabric,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

K>* 1.1.1.1/32 [0/0] is directly connected, dummy1, 00:00:04
K>* 1.1.1.2/32 [0/0] is directly connected, dummy1, 00:00:04
K>* 1.2.1.1/32 [0/0] is directly connected, dummy2, 00:00:04
K>* 1.2.1.2/32 [0/0] is directly connected, dummy2, 00:00:04
S>* 2.1.1.1/32 [1/0] is directly connected, dummy1, weight 1, 00:00:04
S>* 2.1.1.2/32 [1/0] is directly connected, dummy1, weight 1, 00:00:04
S>* 2.2.1.1/32 [1/0] is directly connected, dummy2, weight 1, 00:00:04
S>* 2.2.1.2/32 [1/0] is directly connected, dummy2, weight 1, 00:00:04
C>* 3.1.1.0/24 is directly connected, dummy1, 00:00:04
C>* 3.1.2.0/24 is directly connected, dummy1, 00:00:04
C>* 3.2.1.0/24 is directly connected, dummy2, 00:00:04
C>* 3.2.2.0/24 is directly connected, dummy2, 00:00:04
C>* 192.168.1.0/24 is directly connected, r1-eth0, 00:00:21
r1# show ip nht
r1#
```